pkg/controller/master/storagenetwork/storage_network.go

```diff
@@ -0,0 +1,310 @@
+	longhornSettings     ctllonghornv1.SettingClient
+	longhornSettingCache ctllonghornv1.SettingCache
+	longhornVolumes      ctllonghornv1.VolumeClient
+	longhornVolumeCache  ctllonghornv1.VolumeCache
+	prometheus           ctlmonitoringv1.PrometheusClient
+	prometheusCache      ctlmonitoringv1.PrometheusCache
+	deployments          v1.DeploymentClient
+	deploymentCache      v1.DeploymentCache
+	managedCharts        ctlmgmtv3.ManagedChartClient
+	managedChartCache    ctlmgmtv3.ManagedChartCache
+	settings             ctlharvesterv1.SettingClient
+	settingsController   ctlharvesterv1.SettingController
+}
+
+// register the setting controller and reconsile longhorn setting when storage network changed
+func Register(ctx context.Context, management *config.Management, opts config.Options) error {
+	settings := management.HarvesterFactory.Harvesterhci().V1beta1().Setting()
+	longhornSettings := management.LonghornFactory.Longhorn().V1beta1().Setting()
+	longhornVolumes := management.LonghornFactory.Longhorn().V1beta1().Volume()
+	prometheus := management.MonitoringFactory.Monitoring().V1().Prometheus()
+	deployments := management.AppsFactory.Apps().V1().Deployment()
+	managedCharts := management.RancherManagementFactory.Management().V3().ManagedChart()
+
+	controller := &Handler{
+		ctx:                  ctx,
+		longhornSettings:     longhornSettings,
+		longhornSettingCache: longhornSettings.Cache(),
+		longhornVolumes:      longhornVolumes,
+		longhornVolumeCache:  longhornVolumes.Cache(),
+		prometheus:           prometheus,
+		prometheusCache:      prometheus.Cache(),
+		settings:             settings,
+		settingsController:   settings,
+		deployments:          deployments,
+		deploymentCache:      deployments.Cache(),
+		managedCharts:        managedCharts,
+		managedChartCache:    managedCharts.Cache(),
+	}
+
+	settings.OnChange(ctx, ControllerName, controller.OnStorageNetworkChange)
+	return nil
+}
+
+// webhook needs check if VMs are off
+// webhook needs check if nad is existing
+func (h *Handler) OnStorageNetworkChange(key string, setting *harvesterv1.Setting) (*harvesterv1.Setting, error) {
+	if setting == nil || setting.DeletionTimestamp != nil || setting.Name != settings.StorageNetworkName {
+		return nil, nil
+	}
+
+	if setting.Value == h.getLonghornStorageNetwork() {
+		// check if we need to restart monitoring pods
+		if !h.checkPodStatusAndStart() {
+			h.settingsController.EnqueueAfter(setting.Name, 5*time.Second)
+		}
+
+		// finish
+		return nil, nil
+	}
+
+	logrus.Infof("storage network change:%s", setting.Value)
+
+	// if replica eq 0, skip
+	// save replica to annotation
+	// set replica to 0
+	if !h.checkPodStatusAndStop() {
+		logrus.Infof("Requeue to check pod status again")
+		h.settingsController.EnqueueAfter(setting.Name, 5*time.Second)
+		return nil, nil
+	}
+
+	logrus.Infof("all pods are stopped")
+
+	// check volume detach before put LH settings
+	if !h.checkLonghornVolumeDetach() {
```

Comment:
If the setting is reverted / change back to default, here maybe should avoid check status or skip error

Suppose the change to dedicated network has error, it CAN NOT finish, then user try to reset value back, those check should allow this path.

Reply:
if users want to roll back, we should need to wait for volumes detached too.

Reply:
if users didn't stop all Pods with volumes and rollback to default, it will hit first condition and restart the pods.
because it will not setup LH settings

Reply:
if users' config is synced to LH settings, we have to wait for volume detached again to roll back